/**
 * Copyright 2021 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";
package deployer;
import "var_manager.proto";

service DeployerService {
  rpc DeployerProcess(DeployerRequest) returns (DeployerResponse) {}
}

enum DeployerRequestType {
  kInitRequest = 0;
  kDisconnect = 1;
  kPreDownloadModel = 2;
  kDownloadModel = 3;
  kLoadModel = 4;
  kUnloadModel = 5;
  kPreDeployModel = 6;
  kDownloadVarManager = 7;
  kDownloadSharedContent = 8;
  kDeployRankTable = 9;
  kDownloadConf = 10;
  kHeartbeat = 127;
}

enum DeviceConfigType {
  kLogConfig = 0;
  kDumpConfig = 1;
  kProfilingConfig = 2;
  kCommonConfig = 127;
}

message DeployerRequest {
  int64 client_id = 1;
  DeployerRequestType type = 2;
  oneof body {
    InitRequest init_request = 3;
    PreDownloadModelRequest pre_download_request = 4;
    DownloadModelRequest download_model_request = 5;
    LoadModelRequest load_model_request = 6;
    UnloadModelRequest unload_model_request = 7;
    PreDeployModelRequest pre_deploy_model_request = 8;
    MultiVarManagerRequest multi_var_manager_request = 9;
    SharedContentDescRequest shared_content_desc_request = 10;
    DeployRankTableRequest deploy_rank_table_request = 11;
    DownloadConfigRequest download_config_request = 12;
  }
}

message DeployerResponse {
  uint32 error_code = 1;
  string error_message = 2;
  oneof body {
    InitResponse init_response = 3;
  }
}

message InitRequest {
  string token = 1;
  int32 logic_device_id = 2;
}

message InitResponse {
  int64 client_id = 1;
  string dgw_host_ip = 2;
  int32 dgw_port = 3;
  int32 dev_count = 4;
}

message PreDownloadModelRequest {
  uint32 model_id = 1;
  uint32 root_model_id = 2;
  uint64 total_size = 3;
  int32 device_id = 4;
}

message DownloadModelRequest {
  uint32 model_id = 1;
  uint32 root_model_id = 2;
  uint64 offset = 3;
  bytes om_content = 4;
  int32 device_id = 5;
}

message DownloadConfigRequest {
  DeviceConfigType sub_type = 1;
  int32 device_id = 2;
  bytes config_data = 3;
}

message QueueDesc {
  int32 type = 1;
  string name = 2;
  uint32 depth = 3;
  string host_ip = 4;
  int32 port = 5;
  string key = 6;
  repeated int32 queue_indices = 7;
}

message QueueBinding {
  int32 src_queue_index = 1;
  int32 dst_queue_index = 2;
}

message SubmodelDesc {
  string model_name = 1;
  uint64 model_size = 2;
  repeated int32 input_queue_indices = 3;
  repeated int32 output_queue_indices = 4;
}

message ExchangePlan {
  string master_ip = 1;
  int32 master_port = 2;
  repeated QueueDesc queues = 3;
  repeated QueueBinding bindings = 4;
}

message MultiVarManagerRequest {
  MultiVarManagerInfo multi_var_manager_info = 1;
  int32 device_id = 2;
}

message SharedContentDescRequest {
  SharedContentDescription shared_content_desc = 1;
  int32 device_id = 2;
}

message PreDeployModelRequest {
  uint32 root_model_id = 1;
  ExchangePlan exchange_plan = 2;
  repeated SubmodelDesc submodels = 3;
  int32 device_id = 4;
  int64 route_id = 5;
}

message DeployRankTableRequest {
  int32 device_id = 1;
  string rank_table = 2;
}

message LoadModelRequest {
  uint32 model_id = 1;
  uint32 root_model_id = 2;
  int32 device_id = 3;
  bool is_unknown_shape = 4;
}

message UnloadModelRequest {
  uint32 model_id = 1;
  int32 device_id = 2;
}

enum ExecutorRequestType {
  EXECUTOR_REQUEST_TYPE_PRE_DOWNLOAD = 0;
  EXECUTOR_REQUEST_TYPE_DOWNLOAD = 1;
  EXECUTOR_REQUEST_TYPE_LOAD = 2;
  EXECUTOR_REQUEST_TYPE_UNLOAD = 3;
}

message ExecutorRequest {
  message PreDownloadModelRequest {
    uint32 model_id = 1;
    uint32 root_model_id = 2;
    uint64 model_size = 3;
  }

  message DeployRankTableRequest {
    string rank_table = 1;
    string rank_id = 2;
  }

  message DownloadModelRequest {
    uint32 model_id = 1;
    uint32 root_model_id = 2;
    uint64 offset = 3;
    bytes model_data = 4;
  }

  message LoadModelRequest {
    uint32 model_id = 1;
    uint32 root_model_id = 2;
    repeated uint32 input_queues = 3;
    repeated uint32 output_queues = 4;
  }

  message UnloadModelMessage {
    uint32 model_id = 1;
  }

  oneof body {
    PreDownloadModelRequest pre_download_message = 2;
    DownloadModelRequest download_model_message = 3;
    LoadModelRequest load_model_message = 4;
    UnloadModelMessage unload_model_message = 5;
    MultiVarManagerInfo multi_var_manager_info = 6;
    SharedContentDescription shared_content_desc = 7;
    DeployRankTableRequest deploy_rank_table_message = 8;
  }
}

message ExecutorResponse {
  uint32 status_code = 1;
  string error_message = 2;
}
